src/viewport/index.tsx

```diff
@@ -627,57 +638,98 @@ class Viewport extends React.Component<
-        const timePerFrame = 1000 / framesPerSecond; // the time interval at which to re-render
-        const now = Date.now();
-        const elapsedTime = now - this.lastRenderTime;
-        const totalElapsedTime = now - this.startTime;
-        if (elapsedTime > timePerFrame) {
-            if (simulariumController.isChangingFile) {
-                this.visGeometry.render(totalElapsedTime);
-                this.lastRenderTime = Date.now();
-                this.lastRenderedAgentTime = -1;
-                simulariumController.markFileChangeAsHandled();
-
-                this.animationRequestID = requestAnimationFrame(this.animate);
+        const nextFrame =
+            simulariumController.visData.frameCache.getLastFrameNumber() + 1;
+        simulariumController.visData.gotoNextFrame();
+        simulariumController.resumeStreaming(nextFrame);
+        this.lastFrameAdvanceTime = this.lastRenderTime; // Use last render time instead of 0
+    }
 
-                return;
-            }
-            const currentFrame = visData.currentFrameData;
-            if (currentFrame.time != this.lastRenderedAgentTime) {
-                if (currentFrame.agentCount > 0) {
-                    this.dispatchUpdatedTime({
-                        time: currentFrame.time,
-                        frameNumber: currentFrame.frameNumber,
-                    });
-                    this.visGeometry.update(currentFrame);
-                    this.lastRenderedAgentTime = currentFrame.time;
-                    this.updateFollowObjectData();
-                }
-            }
+    private shouldFrameAdvance(timestamp: number): boolean {
+        const { simulariumController, playbackSpeed } = this.props;
+        const timeSinceLastAdvance =
+            timestamp - this.lastFrameAdvanceTime;
 
-            if (!visData.atLatestFrame() && simulariumController.isPlaying()) {
-                visData.gotoNextFrame();
-            }
-            this.stats.begin();
-            this.visGeometry.render(totalElapsedTime);
-            if (this.recorder?.isRecording) {
-                this.recorder.onFrame();
-            }
-            this.stats.end();
-            this.lastRenderTime = Date.now();
+        return (
+            !simulariumController.visData.atLatestFrame() &&
+            simulariumController.isPlaying() &&
+            timeSinceLastAdvance >= 1000 / playbackSpeed
+        );
+    }
+
+    private renderCurrentFrame(currentFrame: CachedFrame): void {
+        if (
+            currentFrame.time !== this.currentSimulationTime &&
+            currentFrame.agentCount > 0
+        ) {
+            this.currentSimulationTime = currentFrame.time;
+            this.dispatchUpdatedSimulationTime({
+                time: this.currentSimulationTime,
+                frameNumber: currentFrame.frameNumber,
+            });
+            this.visGeometry.update(currentFrame);
+            this.updateFollowObjectData();
+        }
```

Comment:
can you add a note about what this is doing? I haven't looked at these functions in a long time but I'm not clear on why `renderCurrentFrame` would be first checking if the time isn't matching and then doing nothing if it does match. Is there a case where this is called and currentFrame.time === this.currentSimulationTime? 
Maybe it's just the name of this function. I would think `renderCurrentFrame` would always be doing something, not just conditionally. Maybe call this, updateToCurrentFrame?